src/gather.js

```diff
@@ -0,0 +1,59 @@
+'use strict';
+
+import {Tensor, sizeOfShape} from './lib/tensor.js';
+import {validateGatherParams} from './lib/validate-input.js';
+
+/**
+ * Gather values of the input tensor along an axis according to the indices.
+ * @param {Tensor} input
+ * @param {Tensor} indices
+ * @param {MLGatherOptions} [options]
+ * @return {Tensor}
+ */
+export function gather(input, indices, {axis = 0} = {}) {
+  validateGatherParams(...arguments);
+  const shapeInput = input.shape;
+
+  // set outputShape following Spec Algorithm
+  // let dimCount = 0
+  // let rankOutput = 0;
+  // let shapeOutput = [];
+  // for (dimCount = 0; dimCount < shapeInput.length; dimCount++) {
+  //   if (dimCount === axis) {
+  //     break;
+  //   } else {
+  //     shapeOutput[dimCount] = shapeInput[dimCount];
+  //   }
+  // }
+  // rankOutput = dimCount;
+  // for (dimCount = 0; dimCount < indices.shape.length; dimCount++) {
+  //   shapeOutput[rankOutput + dimCount] = indices.shape[dimCount];
+  // }
+  // rankOutput = rankOutput + dimCount;
+  // for (dimCount = 0; dimCount < shapeInput.length; dimCount++) {
+  //   if (dimCount <= axis) {
+  //     continue;
+  //   } else {
+  // here index shoud be set to (rankOutput + dimCount - axis - 1)
+  //     shapeOutput[rankOutput + dimCount - axis - 1] = shapeInput[dimCount];
+  //   }
+  // }
+
+  // optimized set outputShape using JavaScipt slice and concat
+  const shapeOutput = shapeInput.slice(0, axis).concat(indices.shape, shapeInput.slice(axis + 1));
+  const output = new Tensor(shapeOutput);
+
+  for (let i = 0; i < sizeOfShape(shapeOutput); ++i) {
```

Comment:
Consider using another variable name? I suppose the 'i' being used in the following comment is different one.